mirgecom/initializers.py

```diff
@@ -0,0 +1,218 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+import numpy.linalg as la  # noqa
+from pytools.obj_array import (
+    flat_obj_array,
+    make_obj_array,
+)
+import pyopencl.clmath as clmath
+from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
+
+from grudge.symbolic.primitives import TracePair
+from mirgecom.eos import IdealSingleGas
+
+
+class Vortex2D:
+    """Implements the isentropic vortex after
+        - Y.C. Zhou, G.W. Wei / Journal of Computational Physics 189 (2003) 159
+        - JSH/TW Nodal DG Methods, p. 209
+
+    A call to this object after creation/init creates
+    the isentropic vortex solution at a given time (t)
+    relative to the configured origin (center) and
+    background flow velocity (velocity).
+
+    This object also functions as a boundary condition
+    by providing the "get_boundary_flux" routine to
+    prescribe exact field values on the given boundary.
+    """
+
+    def __init__(
+        self,
+        beta=5,
+        center=np.zeros(shape=(2,)),
+        velocity=np.zeros(shape=(2,)),
+    ):
+        self._beta = beta
+        self._center = center  # np.array([5, 0])
+        self._velocity = velocity  # np.array([0, 0])
+
+    def __call__(self, t, x_vec, eos=IdealSingleGas()):
+        # Y.C. Zhou, G.W. Wei / Journal of Computational Physics 189 (2003) 159
+        # also JSH/TW Nodal DG Methods, p. 209
+        vortex_loc = self._center + t * self._velocity
+
+        # coordinates relative to vortex center
+        x_rel = x_vec[0] - vortex_loc[0]
+        y_rel = x_vec[1] - vortex_loc[1]
+
+        gamma = eos.gamma()
+        r = clmath.sqrt(x_rel ** 2 + y_rel ** 2)
+        expterm = self._beta * clmath.exp(1 - r ** 2)
+        u = self._velocity[0] - expterm * y_rel / (2 * np.pi)
+        v = self._velocity[1] + expterm * x_rel / (2 * np.pi)
+        mass = (
+            1 - (gamma - 1) / (16 * gamma * np.pi ** 2) * expterm ** 2
+        ) ** (1 / (gamma - 1))
+        p = mass ** gamma
+
+        e = p / (gamma - 1) + mass / 2 * (u ** 2 + v ** 2)
+
+        return flat_obj_array(mass, e, mass * u, mass * v)
+
+    def get_boundary_flux(
+        self, discr, w, t=0, btag=BTAG_ALL, eos=IdealSingleGas()
+    ):
+        queue = w[0].queue
+
+        # help - how to make it just the boundary nodes?
+        nodes = discr.nodes().with_queue(queue)
+        vortex_soln = self.__call__(t, nodes)
+        dir_bc = discr.interp("vol", btag, vortex_soln)
+        dir_soln = discr.interp("vol", btag, w)
+        from mirgecom.euler import _facial_flux  # hrm
+
+        return _facial_flux(
+            discr, w_tpair=TracePair(btag, dir_soln, dir_bc), eos=eos,
+        )
```

Comment:
It shouldn't.  This smelled bad as it was written.  We do need whatever boundary handler to know how to call the construct that is providing the exact solution.   These can vary, and have different sorts of input needs.   Please see if what I have done with it using a "PrescribedBoundary" type is ok for you.

Reply:
It still contains interpolation code, which I don't like. Generally, I think it's fine to supply all of them with all inputs. But, but, but, I hear you say, what if they don't use all these inputs? Two answers:

* Once we're on lazy eval, it doesn't matter. You don't pay for computations whose results you don't use.
* If you'd like to make it happy until then, you could pass in lambdas that the function can call to trigger the computation in question.

Reply:
Initializers no longer have this.  I will close this convo up so that we can re-center this discussion around the relevant construct (BC) in boundary.py.   The Boundary (still a) class method now returns a TracePair.  I suspect it still isn't exactly how you would like it to be.